feat: additional template functions

matkam opened the pull request:
**Description**

Adds a few useful text/template functions:

- replace: replace all substrings in a target string
- isIPv6: exposes an existing function that returns a bool to let you know if a string is an IPv6 address
- isIPv4: a new function that returns a bool to let you know if a string is an IPv4 address

<!-- Please link to all GitHub issue that this pull request implements(i.e. Fixes #123) -->
N/A

**Checklist**

- [x] Unit tests updated
- [x] End user documentation updated

johngmyers commented:
Please submit a CLA per the instructions above.

matkam commented:
> Please submit a CLA per the instructions above.

Yes 👍 
I'm just waiting on my employer to click the button.

matkam commented:
@johngmyers we are good to go here 👍

johngmyers commented on source/source.go:
```diff
@@ -390,3 +393,8 @@ func isIPv6String(ip string) bool {
+
+func isIPv4String(input string) bool {
+	netIP := net.ParseIP(input)
+	return netIP != nil && netIP.To4() != nil
```
Is this function intended to return `true` for IPv4-mapped IPv6 addresses?

matkam replied:
It just returns `true` when the input string is an IPv4 address. It'll return `false` if its IPv6 or otherwise invalid IP address.

johngmyers replied:
https://go.dev/play/p/2SNDvrwxdh-

matkam replied:
how about something like this?
```
func isIPv4String(input string) bool {
	netIP := net.ParseIP(input)
	return netIP != nil && netIP.To4() != nil && !strings.Contains(input, ":")
}
```

johngmyers commented:
Sprig has `replace`, which is similar. Perhaps we should follow their syntax?

johngmyers commented:
/ok-to-test

matkam commented:
> Sprig has `replace`, which is similar. Perhaps we should follow their syntax?

does `replace` work differently? I could also add `strings.replace` to the `FuncMap`.

johngmyers commented:
I believe `replace` has the arguments in a different order. See the documentation in https://masterminds.github.io/sprig/strings.html

matkam commented:
> I believe `replace` has the arguments in a different order. See the documentation in https://masterminds.github.io/sprig/strings.html

sprig looks like a nice package. how about just importing it and using `hermeticTxtFuncMap` like this:
```
func parseTemplate(fqdnTemplate string) (tmpl *template.Template, err error) {
	if fqdnTemplate == "" {
		return nil, nil
	}
	
	funcs := sprig.HermeticTxtFuncMap()
	funcs["isIPv6"] = isIPv6String
	funcs["isIPv4"] = isIPv4String

	return template.New("endpoint").Funcs(funcs).Parse(fqdnTemplate)
}
```

matkam commented:
@johngmyers let me know what you think. I've updated the template functions to use Sprig's hermetic text functions, and fixed the `isIPv4String` function so that it only returns `true` for valid IPv4 strings.

mloiseleur commented on docs/tutorials/nodes.md:
```diff
@@ -6,6 +6,7 @@ Using nodes (`--source=node`) as source is possible to synchronize a DNS zone wi
 The node source adds an `A` record per each node `externalIP` (if not found, any IPv4 `internalIP` is used instead).
 It also adds an `AAAA` record per each node IPv6 `internalIP`.
 The TTL of the records can be set with the `external-dns.alpha.kubernetes.io/ttl` node annotation.
+The FQDN template includes a function to replace all `.` with `-` in the node name, which is useful for cloud providers that include dots in the node name.
```
```suggestion
The FQDN template provides more than 100+ functions, documented [here](https://go-task.github.io/slim-sprig/). For instance, it includes a function to replace all `.` with `-` in the node name, which can be useful with cloud providers that include dots in the node name.
```

mloiseleur commented:
@matkam Wdyt about adding `isIPv4` and `isIPv6` directly into slim-sprit ?

matkam commented:
@mloiseleur https://github.com/go-task/slim-sprig/pull/16

mloiseleur commented:
/hold for https://github.com/go-task/slim-sprig/pull/16

mloiseleur commented:
@matkam Since slim-sprig maintainer does not review your PR, I think we can merge this one. Do you think you can rebase this PR ? And precise in documentation those additional functions of `isIPv4` and `isIPv6` ?

matkam commented:
@mloiseleur done!

mloiseleur commented:
/lgtm

mloiseleur commented:
A review from @szuecs or @Raffo 
They will review this PR when they have time.

mloiseleur commented:
/unhold

mloiseleur commented:
/retitle feat: additional template functions

Jasper0147Ai commented:
I can't link the two account togetherOn May 22, 2024 7:31 AM, Simon Ostendorf ***@***.***> wrote:
Hello, I would like to see this merged and released. What needs to be done?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

szuecs commented on source/source.go:
```diff
@@ -29,6 +29,7 @@ import (
 	"time"
 	"unicode"
 
+	sprig "github.com/go-task/slim-sprig"
```
Please no import for a single func.

matkam replied:
@szuecs it includes several useful go template functions: https://github.com/go-task/slim-sprig/blob/master/functions.go

Would you rather not include these functions, and only keeping `isIPv4`?

matkam commented:
/remove-lifecycle stale

ivankatliarchuk commented on go.mod:
```diff
@@ -37,6 +37,7 @@ require (
 	github.com/ffledgling/pdns-go v0.0.0-20180219074714-524e7daccd99
 	github.com/go-gandi/go-gandi v0.7.0
 	github.com/go-logr/logr v1.4.2
+	github.com/go-task/slim-sprig/v3 v3.0.0
```
the only concern is that the libary had last release 2+ years ago in 2023. Could you justify/convince the library maintainer to 

- refresh a release tag 
- refresh go.mod with go.sum
- would be nice to sync `go` version as well, but not critical

have you checked this library for known vulnerabilities?

```
trivy repo --scanners vuln .
```

or with snyk open source checkers

ivankatliarchuk replied:
This actually quite tricky. For example github.com/Masterminds/sprig is heavier, but it has 4k stars and almost ~100 contributors.... So worth to share a comparison on why this library vs upstream one

matkam replied:
It doesn't make a difference for my use case. I only need the `replace` function. Context [here](https://github.com/kubernetes-sigs/external-dns/pull/3949#issuecomment-1736469058). Would you prefer importing github.com/Masterminds/sprig instead, or even creating a similar function here and avoiding the import altogether?

matkam replied:
It doesn't make a difference for my use case. I only need the `replace` function. Context [here](https://github.com/kubernetes-sigs/external-dns/pull/3949#issuecomment-1736469058). Would you prefer importing github.com/Masterminds/sprig instead, or even creating a similar function here and avoiding the import altogether?

ivankatliarchuk replied:
I have no opinion. If you refactor logic as per review, and cover functions with tests, we will decide on a library at the very end. Both libraries have they advantages. Or as you mention, we may not need any libraries at this time, as only few functions required.

ivankatliarchuk replied:
If you only need one function - so no import required. Better to creat own function. It's more about compatibility, we should have tests for all the functions provided.

ivankatliarchuk commented on source/source.go:
```diff
@@ -266,6 +269,25 @@ func suitableType(target string) string {
+// isIPv6String reports whether the target string is an IPv6 address,
+// including IPv4-mapped IPv6 addresses.
+func isIPv6String(target string) bool {
+	netIP, err := netip.ParseAddr(target)
```
could you share code coveragate for this functions?

Can we have specific tests just in source_test.go

I know it's a private function, but we may move them at some point to utils.go, this will help with maintenance

Worth to add BenchmarkTest... as well, and if you could, share results

ivankatliarchuk commented on source/source.go:
```diff
@@ -266,6 +269,25 @@ func suitableType(target string) string {
+}
+
+// isIPv4String reports whether the target string is an IPv4 address.
+func isIPv4String(target string) bool {
```
same as other function

ivankatliarchuk commented on docs/sources/nodes.md:
```diff
@@ -111,7 +112,7 @@ spec:
         - --aws-zone-type=public
         - --registry=txt
-        - --fqdn-template={{.Name}}.external-dns-test.my-org.com
+        - --fqdn-template={{.Name | replace "." "-"}}.external-dns-test.my-org.com
```
I never tried this `fqdn-template`, could you create a page or a section with bit more information about this `fqdn-template` and it's use cases and advantages?

I can see an example for raw kubernetes manifest. How this behave in helm chart? as the syntax is very similar. I would assume helm generation will fail, am I correct?

ivankatliarchuk reviewed (changes requested):
Hi @matkam. Overall looks ok, if you still have time and interest to add this to a master branch, I've shared few bits that worth addressing.

Could you also provide a set of manual test steps using manifests and kubectl commands? This will help in understanding better the change and verifying the implementation.

/assign

ivankatliarchuk commented:
May I also ask to write tests similar to like helm does as well https://github.com/helm/helm/blob/0d66425d9a745d8a289b1a5ebb6ccc744436da95/pkg/engine/funcs_test.go#L27

We probably should have same structure for files as helm too, aka move this functions to `template_functions.go` or smth as agree, this is quite a powerful concept

ivankatliarchuk commented on docs/sources/nodes.md:
```diff
@@ -6,6 +6,7 @@ Using nodes (`--source=node`) as source is possible to synchronize a DNS zone wi
 The node source adds an `A` record per each node `externalIP` (if not found, any IPv4 `internalIP` is used instead).
 It also adds an `AAAA` record per each node IPv6 `internalIP`.
 The TTL of the records can be set with the `external-dns.alpha.kubernetes.io/ttl` node annotation.
+The FQDN template provides more than 100+ functions, documented [here](https://go-task.github.io/slim-sprig/). For instance, it includes a function to replace all `.` with `-` in the node name, which can be useful with cloud providers that include dots in the node name. There are two additional functions available on top of the standard sprig functions: `isIPv4` and `isIPv6`. The functions can be ...
```
could we also have a markdown table with all available functions, description and short  usage example?

ivankatliarchuk commented on source/source.go:
```diff
@@ -147,9 +148,11 @@ func parseTemplate(fqdnTemplate string) (tmpl *template.Template, err error) {
-		"trimPrefix": strings.TrimPrefix,
-	}
+
+	funcs := sprig.HermeticTxtFuncMap()
```
could we use go template instead here?

extra := template.FuncMap{
    "isIPv6" : isIPv6String
    "isIPv4" : isIPv4String
}

and add them to `funcs` if it's possible

ivankatliarchuk replied:
this code most likely better to move to 
```
func funcMap() template.FuncMap {}
```

so we could have an isolated code coverage for multiple functions

matkam commented:
> Hi @matkam. Overall looks ok, if you still have time and interest to add this to a master branch, I've shared few bits that worth addressing.

Hi @ivankatliarchuk, thanks for looking at this PR. I definitely do want to get it merged in. I will look for some time in the coming weeks to address your review items.

ivankatliarchuk commented:
/label tide/merge-method-squash

mloiseleur commented:
now that https://github.com/kubernetes-sigs/external-dns/pull/5354 has been merged, the path for this PR should be simpler.

@matkam Do you think you can rebase your PR and address review comments from ivan ?

matkam commented:
> @matkam Do you think you can rebase this PR and address review comments from ivan ?

@mloiseleur @ivankatliarchuk this PR is ready for your review.

ivankatliarchuk reviewed (changes requested):
Nice one.

Any chance to add tests fro specific functions? I know most of them tested, but go not able to identify that the function is under test when tested over templating

![Screenshot 2025-05-11 at 08 34 44](https://github.com/user-attachments/assets/c7701fbe-3fde-428a-ace6-c273da4bc428)

ivankatliarchuk commented:
👋 @mloiseleur I'm confident about this change. If you ok, we could merge this. I'll bump coverage myself as follow-up

/lgtm

matkam commented:
> Any chance to add tests fro specific functions? I know most of them tested, but golang coverage is not able to identify that the function is under test when tested over templating

@ivankatliarchuk I've added coverage for the functions:
```
> go test -cover ./source/fqdn/...
ok      sigs.k8s.io/external-dns/source/fqdn    0.175s  coverage: 100.0% of statements
```

ivankatliarchuk commented:
Nice one /lgtm

ivankatliarchuk reviewed (approved):
lgtm

mloiseleur commented:
It took 2 years, but we made it, finally. 
Thanks @matkam 
/approve
